HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,85 @@
+def update_child_links(node, dist, parent, neighbors, level):
+    '''
+    Filter nodes whose distances are the current level and return
+    all links of child nodes with updated distance
+    Otherwises, return the original one
+    '''
+    if dist == level:
+        result = [(node, (dist, parent, neighbors))]
+        result = result + [(neighbor, (level+1, node, [])) for neighbor in neighbors]
+        return result
+    else:
+        return [(node, (dist, parent, neighbors))]
+
+def update_graph(x, y):
+    '''
+    Choose the node with the minimum distance while keeping all
+    neighbors of each node
+    '''
+    dist_x = x[0]
+    dist_y = y[0]
+    parent_x = x[1]
+    parent_y = y[1]
+    neighbors_x = x[2]
+    neighbors_y = y[2]
+
+    if neighbors_x == []:
+        neighbors = neighbors_y
+    elif neighbors_y == []:
+        neighbors = neighbors_x
+    elif (neighbors_x == []) and (neighbors_y == []):
+        neighbors = []
+
+    if dist_x < dist_y:
+        return (dist_x, parent_x,neighbors)
+    else:
+        return (dist_y, parent_y,neighbors)
+
+def BFS_with_short_path(rdd_graph, start_node, end_node):
+    '''
+    Given a (K, V) RDD with K = node and V = [connecting nodes],
+    find the shortest paths from a source node to all other nodes in a graph.
+    '''
+    # rdd_graph => (node, (distance, parent, neighbors)) RDD
+    rdd_graph = rdd_graph.map(lambda (node, neighbors): (node, (0, None, neighbors)) if (node == start_node) else (node, (float('inf'), None, neighbors)))
+    level = 0
+    while True:
+      # flatMap graph to RDD 
+      rdd_graph = rdd_graph.flatMap(lambda (node, (dist, parent, neighbors)) : update_child_links(node, dist, parent, neighbors, level))
+      rdd_graph = rdd_graph.reduceByKey(lambda x, y: update_graph(x,y))
+
+      # if distance of end_node is not infinite number
+      if rdd_graph.map(lambda x: x).lookup(end_node)[0][0] != float('inf'):
+        break
+      level += 1
+
+    level += 1
```

Comment:
Actually, it won't do much of a difference as the accumulator is used just for storing the number of iterations.
If the accumulator was used to count the number of nodes with a new distance value (during a map or a reduce phase), it would have been useful.